pageserver/src/http/models.rs

```diff
@@ -1,20 +1,121 @@
+use crate::timelines::TimelineInfo;
+use anyhow::{anyhow, bail, Context};
 use serde::{Deserialize, Serialize};
-
-use crate::ZTenantId;
-use zenith_utils::zid::ZNodeId;
+use zenith_utils::{
+    lsn::Lsn,
+    zid::{HexZTenantId, HexZTimelineId, ZNodeId, ZTenantId, ZTimelineId},
+};
 
 #[derive(Serialize, Deserialize)]
-pub struct BranchCreateRequest {
-    #[serde(with = "hex")]
-    pub tenant_id: ZTenantId,
-    pub name: String,
-    pub start_point: String,
+pub struct TimelineCreateRequest {
+    pub new_timeline_id: Option<HexZTimelineId>,
+    pub ancestor_timeline_id: Option<HexZTimelineId>,
+    pub ancestor_start_lsn: Option<Lsn>,
 }
 
 #[derive(Serialize, Deserialize)]
 pub struct TenantCreateRequest {
+    pub new_tenant_id: Option<HexZTenantId>,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct TimelineInfoResponse {
+    pub kind: String,
+    #[serde(with = "hex")]
+    timeline_id: ZTimelineId,
```

Comment:
We can use both, I personally find this variant slightly better since no `From` conversion is done in regular code this way.

I guess, it would make more sense to have hex variants everywhere, if we would have a bit more explicit API around "serializeable json objects" in our logic (`models.rs` is a step in that direction, but not yet complete alas).

Reply:
I still do not have complete picture in my head regarding this serde related stuff. I think it may not be possible to use hex variants everywhere, if some struct is serialized to both, binary and json we'll have to duplicate the struct with a Hex id variand and use From impl to convert one to another

Reply:
There are 3 things happening more or less separately:
* `#[derive(Serialize, Deserialize)]` that does conventional (de)serialization
* `#[serde(with = "smth")]` attribute that can be placed on any field (in this context), overriding its (de)serialize entirely, it's not even required for that annotated field to have `(De)Serialize` trait derived
* different newtypes with different `(De)Serialize` impls.

So

```rust
#[derive(Deserialize, Serialize)]
struct Foo {
    one: ZTimeineId,
    #[serde(with = "hex")]
    two: ZTimeineId,
    three: ZHexTimeineId,
}

fn main() {
    dbg!(serde_json::to_string(&Foo {...}));
}
```

will print `one` as an array of bytes dues its serialization rules (`derive` on a newtype over byte array), will take an override for `two` and use `hex` module to serialize it as a hex string, and `three` as a hex string due its custom `(De)Serialize` impls (as a hex string, with visitor)

Reply:
Hm, having both Hex id newtypes and serde(with="hex") attribute looks a bit complex and confusing but I dont know what we can do about that

Reply:
Yes, neither solution is perfect alas, I've tried to find a way to make (de)serialization generic but that did not work, converting everything into ZHex* variant forces you to write extra bloat code, mixed annotations and Hex newtypes are confusing but seems like a lesser evil given that person reads the ZHex* doc.